src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -0,0 +1,152 @@
+        return segment_position.value() < search_value;
+      });
+    }
+  }
+
+  // This function sets the offset(s) which delimit the result set based on the predicate condition and the sort order
+  void _set_begin_and_end() {
+    if (_predicate_condition == PredicateCondition::Equals) {
+      _begin = _get_first_bound();
+      _end = _get_last_bound();
+      return;
+    }
+
+    // clang-format off
+    if (_is_ascending) {
+      switch (_predicate_condition) {
+        case PredicateCondition::GreaterThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::GreaterThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::LessThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::LessThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    } else {
+      switch (_predicate_condition) {
+        case PredicateCondition::LessThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::LessThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::GreaterThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    }
+    // clang-format on
+  }
+
+  /*
+   * NotEquals may result in two matching ranges (one below and one above the search_value) and needs special handling.
+   * The function contains three early outs. These are all only for performance reasons and, if removed, would not
+   * change the functionality.
+   */
+  template <typename Functor>
+  void _handle_not_equals(const Functor& functor) {
+    const auto first_occurrence = _get_first_bound();
+    if (first_occurrence == _end) {
+      // If the value isn't found anywhere, call the functor on the whole range and skip the call to _get_last_bound().
+      functor(_begin, _end);
+      return;
+    }
+
+    const auto last_occurrence = _get_last_bound();
+    if (last_occurrence == _end) {
+      // If the last occurrence of the value is right at the end, call the functor from start to first occurrence and
+      // skip the need for boost::join().
+      functor(_begin, first_occurrence);
+      return;
+    }
+
+    if (first_occurrence == _begin) {
+      // If the first occurrence of the value is right at the start, call the functor from last occurrence to end and
+      // skip the need for boost::join().
+      functor(last_occurrence, _end);
+      return;
+    }
+
+    const auto range = boost::join(boost::make_iterator_range(_begin, first_occurrence),
+                                   boost::make_iterator_range(last_occurrence, _end));
+    functor(range.begin(), range.end());
+  }
+
+ public:
+  template <typename Functor>
```

Comment:
This could be more specific. `ResultAdder`?

Reply:
I settled for `ResultConsumer` now :+1: